File: java-to-go-projects/19-graceful-server/Handler.java
package graceful;

import com.sun.net.httpserver.*;
import java.io.*;

/**
 * Application-level HTTP handlers.
 * These are the actual business logic endpoints.
 */
public class Handler {

    /**
     * Health check endpoint — returns 200 with status.
     */
    public static void health(HttpExchange exchange) throws IOException {
        String body = "{\"status\":\"healthy\"}";
        exchange.getResponseHeaders().add("Content-Type", "application/json");
        exchange.sendResponseHeaders(200, body.length());
        exchange.getResponseBody().write(body.getBytes());
        exchange.getResponseBody().close();
    }

    /**
     * Simulates a slow endpoint (for testing graceful shutdown with in-flight requests).
     */
    public static void slow(HttpExchange exchange) throws IOException {
        try {
            Thread.sleep(5000); // simulate slow work
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        String body = "{\"message\":\"slow response complete\"}";
        exchange.getResponseHeaders().add("Content-Type", "application/json");
        exchange.sendResponseHeaders(200, body.length());
        exchange.getResponseBody().write(body.getBytes());
        exchange.getResponseBody().close();
    }

    /**
     * Echo endpoint — returns request info.
     */
    public static void echo(HttpExchange exchange) throws IOException {
        String method = exchange.getRequestMethod();
        String path = exchange.getRequestURI().toString();
        String body = String.format("{\"method\":\"%s\",\"path\":\"%s\"}", method, path);
        exchange.getResponseHeaders().add("Content-Type", "application/json");
        exchange.sendResponseHeaders(200, body.length());
        exchange.getResponseBody().write(body.getBytes());
        exchange.getResponseBody().close();
    }
}

File: java-to-go-projects/19-graceful-server/rust/src/handler.rs
/// GET /health — return 200 with JSON {"status": "ok"}
pub async fn health() -> (u16, String) {
    todo!()
}

/// GET /slow — simulate slow operation (5s sleep), then 200
pub async fn slow() -> (u16, String) {
    todo!()
}

/// ANY /echo — return method + path as response body
pub async fn echo(method: &str, path: &str) -> (u16, String) {
    todo!()
}

/// Route a request to the appropriate handler.
pub async fn route(method: &str, path: &str) -> (u16, String) {
    todo!()
}
